### PR TITLE
sd-device: expose sd_device_get_device_id() and sd_device_get_driver_subsystem()

### PR DESCRIPTION
**Pull request by yuwata**
Also, make sd_device_new_from_subsystem_sysname() stricter as the pair of subsystem and sysname is not unique in general.

### CODE REVIEW

File: src/libsystemd/sd-device/sd-device.c [hunk 1]
```diff
@@ -1658,6 +1672,7 @@ int device_get_device_id(sd_device *device, const char **ret) {
                         if (r == O_DIRECTORY)
                                 return -EINVAL;
 
+                        r = sd_device_get_driver_subsystem(device, &driver_subsystem);
```
Comment (daandemeyer) [inline src/libsystemd/sd-device/sd-device.c]: Hmm querying but never using?

File: src/libsystemd/sd-device/test-sd-device.c [hunk 2]
```diff
@@ -97,12 +99,12 @@ static void test_sd_device_one(sd_device *d) {
...
                                                  "Maybe running on a non-host network namespace.", id);
-                else {
-                        assert_se(r >= 0);
+                else if (r >= 0) {
                         assert_se(sd_device_get_syspath(dev, &val) >= 0);
                         assert_se(streq(syspath, val));
                         dev = sd_device_unref(dev);
-                }
+                } else
+                        ASSERT_EQ(r, -ETOOMANYREFS);
```
Comment (daandemeyer) [inline src/libsystemd/sd-device/test-sd-device.c]: `ASSERT_ERROR()`

File: src/libsystemd/sd-device/test-sd-device.c [hunk 3]
```diff
@@ -85,10 +83,14 @@ static void test_sd_device_one(sd_device *d) {
...
-                assert_se(streq(syspath, val));
-                dev = sd_device_unref(dev);
+
+                r = sd_device_new_from_subsystem_sysname(&dev, subsystem, name);
+                if (r >= 0) {
+                        assert_se(sd_device_get_syspath(dev, &val) >= 0);
+                        assert_se(streq(syspath, val));
+                        dev = sd_device_unref(dev);
+                } else
+                        ASSERT_EQ(r, -ETOOMANYREFS);
```
Comment (daandemeyer) [inline src/libsystemd/sd-device/test-sd-device.c]: `ASSERT_ERROR()`

File: src/libsystemd/sd-device/sd-device.c [hunk 4]
```diff
@@ -345,9 +345,11 @@ _public_ int sd_device_new_from_ifname(sd_device **ret, const char *ifname) {
...
         assert_return(ifname, -EINVAL);
 
-        r = device_new_from_main_ifname(ret, ifname);
-        if (r >= 0)
-                return r;
+        if (ifname_valid(ifname)) {
+                r = device_new_from_main_ifname(ret, ifname);
+                if (r >= 0)
+                        return r;
+        }
```
Comment (poettering) [inline src/libsystemd/sd-device/sd-device.c:352]: (maybe the lesson to learn here, is to switch device_new_from_main_ifname() to use path_join(), but doesn't matter)

File: src/libsystemd/libsystemd.sym [hunk 5]
```diff
@@ -1046,6 +1046,7 @@ global:
         sd_varlink_take_fd;
         sd_varlink_unref;
         sd_varlink_wait;
+        sd_device_get_driver_subsystem;
```
Comment (poettering) [inline src/libsystemd/libsystemd.sym:1050]: hmm, the various sd_device_get_xyz() calls all have a man page. Any chance we can get this one added there, right away too?

File: src/libsystemd/sd-device/sd-device.c [hunk 6]
```diff
@@ -1637,9 +1637,8 @@ static int handle_db_line(sd_device *device, char key, const char *value) {
         }
 }
 
-int device_get_device_id(sd_device *device, const char **ret) {
-        assert(device);
-        assert(ret);
+_public_ int sd_device_get_device_id(sd_device *device, const char **ret) {
```
Comment (poettering) [inline src/libsystemd/sd-device/sd-device.c:1688]: this one too, please add the sd_device_get_xyz() man page with this

File: src/libsystemd/sd-device/sd-device.c (hunk 1 above)
Comment (yuwata) [reply src/libsystemd/sd-device/sd-device.c]: Ouch. I forgot to drop the change.

File: src/libsystemd/sd-device/sd-device.c (hunk 4 above)
Comment (yuwata) [reply src/libsystemd/sd-device/sd-device.c:352]: Yeah, I really dislike `strjoina()`. But in this case, it should be OK.

**Comment by yuwata** [comment]
@DaanDeMeyer and @poettering Thank you for the review.
- Updated man page,
- Added the last commit for udevadm.

PTAL.

File: man/sd_device_get_syspath.xml [hunk 7]
```diff
@@ -171,6 +178,20 @@
     the device name changing, and is relevant for block devices encapsulating devices with changing media
     (e.g. floppy or CD-ROM), or loopback block devices. Only defined for block devices, i.e. those of
     subsystem <literal>block</literal>.</para>
+
+    <para><function>sd_device_get_device_id()</function> returns the short string that identifies the device
+    record. When a block or character device is specified, which has corresponding device node, this returns
+    <literal>b</literal> or <literal>c</literal>, respectively, followed by the device node major and minor
+    numbers separated with a colon. Example: <literal>b259:1</literal> or <literal>c10:121</literal>.
+    When a network interface device is specified, this returns <literal>n</literal> followed by the interface
+    index, which can be obtained by <function>sd_device_get_ifindex()</function>. Eample: <literal>n1</literal>.
```
Comment (poettering) [inline man/sd_device_get_syspath.xml]: Eample → Example

File: man/sd_device_get_syspath.xml [hunk 8]
```diff
@@ -171,6 +178,20 @@
...
+    When a device in the <literal>driver</literal> subsystem is specified, this returns
+    <literal>+drivers:</literal> followed by its driver subsystem and sysfs name separated with a colon.
+    Example: <literal>+drivers:pci:iwlwifi</literal> for a driver device record whose driver subsystem is
+    <literal>pci</literal> and sysfs name is <literal>iwlwifi</literal>,
+    When an other type of device is specified, this function returns <literal>+</literal> followed by its
+    subsystem and sysfs name separated with a colon. Example: <literal>+acpi:ACPI0003:00</literal>,
+    <literal>+input:input16</literal>, or <literal>+pci:0000:00:1f.6</literal>.</para>
```
Comment (poettering) [inline man/sd_device_get_syspath.xml:196]: maybe also add a quick reference to sd_device_new_from_device_id(), pointing out that it converts such a device ID back to an sd_device object.

**Pull request merged**